Custom armor texture templates

Goldorion opened the pull request:
This PR moves the texture templates in plugins (default templates have been moved inside `mcreator-core` ). It also gives us the possibility to import our own armor templates, the same way we can do with templates for the drawing tool. Each texture has its own file. The file also has to finish with something (like "1" for the layer_1). The name of the template (inside the dropdown list) is defined by the name of the file without the addition. 

Here is an example with a template named Plugin:
The name displayed will be "Plugin".
Plugin1 = layer_1
Plugin2 = layer_2
PluginH = Helmet item
PluginBy = Body item
PluginL = Leggings item
PluginBs = Boots item

When a file is missing, I implemented a small checker returning the file of the Standard template. So instead, to return an error and not be able to use the template, it will generate files. If the user only wants his files, he will only have to delete them after.
Here is an example with the explanations (if you can read them XD):
![Capture](https://user-images.githubusercontent.com/38427877/106371969-f4b8c400-6338-11eb-8f4d-92318cbb3ba4.PNG)

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -0,0 +1,82 @@
+		ImageIO.setUseCache(true);
+
+		CACHE.forEach((string, imageIcon) -> {
+			String name = string;
```
No need for re-reference, name original field name

KlemenDEV replied:
No need for re-reference, name original field name

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -0,0 +1,82 @@
+	private static final String MCR_FOLDER = System.getProperty("user.home")+ "\\.mcreator\\";
+
+	public static final Map<String, ImageIcon> CACHE = new ConcurrentHashMap<>();
+	public static final List<String> NAMES = new ArrayList<>();
```
No need for this, you can get key or value set from map directly. Also, make getters for these.

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -20,63 +20,48 @@
 			}
 		});
 		ImageIO.setUseCache(true);
-
```
I think it was ok to cache this part too. Any reason you moved this out into armor maker?

Goldorion replied:
I thought it was what you meant.

KlemenDEV replied:
I meant there was no need for a separate NAMES field, as you could obtain them from the Map. But looking now, you couldn't as the lists differ. It was ok before, sorry for this confusion.

But consider making it Set instead of List if you don't want duplicate values, so you don't need to do contains check.

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -0,0 +1,84 @@
+					name = name.replace(folder + "\\", "");
+				}
+			}
+			NAMES.add(name);
```
Consider having a local set and then convert it to an array, so you don't convert it to an array each time the armor maker is open.

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -0,0 +1,84 @@
+			String name = string;
+			name = name.replace("templates/textures/armormaker/", "");
+			name = name.replace("/", "");
+			String[] types = new String[] { "Bs", "1", "2", "By", "H", "L" };
```
There is no mechanism to not include armor templates in case parts of the template are missing

Goldorion replied:
There is one method to use the Standard template. Check `ArmorImageMakerView#getTextureTemplate`, but I guess I can move it to this method.

KlemenDEV replied:
It is probably better to assemble the whole cache and then use it, after all, that is the point of cache ;)

KlemenDEV reviewed (changes requested):
I now did a full review and noticed some more improvements that could be made

KlemenDEV commented on src/main/java/net/mcreator/ui/init/ArmorMakerTexturesCache.java:
```diff
@@ -0,0 +1,84 @@
+	public static ImageIcon getIcon(@Nullable String itemName) {
+		String fullItemName = itemName + ".png";
+		String folder = "templates/textures/armormaker/";
+		if (itemName != null && CACHE.get(MCR_FOLDER + folder.replace("/", "\\") + fullItemName) != null)
```
You should not be looking up like this, store elements in CACHE by their itemName so you can get them directly so a lookup like this is not needed.

Also hardcoding `MCR_FOLDER = System.getProperty("user.home") + "\\.mcreator\\";` is not ok, but you won't need that anyway if you change CACHE to store names only.

KlemenDEV commented:
@Goldorion I have done some changes to improve code style and simplified some parts.

As extensions such as "L"could appear in the name, I have renamed the files to have a bit more explicit suffixes.

Please verify if all my changes seem ok (possibly test with your custom templates?) and then I can merge the PR.

Goldorion commented:
> 
> 
> @Goldorion I have done some changes to improve code style and simplified some parts.
> 
> As extensions such as "L" could appear in the name, I have renamed the files to have a bit more explicit suffixes.
> 
> Please verify if all my changes seem ok (possibly test with your custom templates?) and then I can merge the PR.

Everything works fine on my side.

KlemenDEV reviewed (approved):
Waiting for tests to finish, otherwise ok

KlemenDEV commented:
@Goldorion do you have any nice demo armors for demo for social media? :)

Goldorion commented:
> 
> 
> @Goldorion do you have any nice demo armors for demo for social media? :)

I'm not really good to draw something, but I will try.

KlemenDEV commented:
> I'm not really good to draw something, but I will try.

I meant if you had any ready, no need to draw if you don't have at hand ;)

Goldorion commented:
> 
> 
> > I'm not really good to draw something, but I will try.
> 
> I meant if you had any ready, no need to draw if you don't have at hand ;)
I won't, I will take textures of Minecraft. ;)

Goldorion commented:
Two screenshots, take the one you prefer.
![Capture2](https://user-images.githubusercontent.com/38427877/107268802-ad0bf980-6a16-11eb-992a-15cd977d0235.PNG)
![Capture](https://user-images.githubusercontent.com/38427877/107268875-c57c1400-6a16-11eb-8313-aee0987a8b16.PNG)

Max094Reikeb reviewed (approved):
Looks good. Ready to be merged.
